### PR TITLE
Refactor/avoid importing kitsune to client

### PR DESCRIPTION
**Pull request by mattyg**
### Summary
I was trying to build the rust client in a docker container and realized I needed go, because it was depending on tx5. There's no reason the rust client should need tx5. This is the prep work to remove it as a dependency -- mostly by removing unnecessary dependencies from `holochain_conductor_api`. I thought it would be simple but it was not -- I'm sure you're aware there is quite the rats nest of dependencies, mostly around kitsune_p2p.

- adds a crate `holochain_nonce`, moves nonce function & type from `holochain_state` into it
- adds a crate `holochain_secure_primitive` (I think this maybe belongs in `holochain_seriailization`)
- adds a crate `holochain_state_types` and extracts some types from `holochain_state` into it
- extract `kitsune_p2p` config & fetch pool types from `kitsune_p2p` into `kitsune_p2p_types`

### TODO:
- [ ] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

**Comment by ThetaSinner** [comment]
Awesome 🎆 Thanks for doing this!

Yes the dependencies around Kitsune need work. `mdns` is another example of a crate I'd like to isolate a bit because it pulls in async-std.

It'll take a bit of time for me to walk through these changes but I'll try and get through it in the morning tomorrow, and I'll add a second reviewer too :)

**Comment by mattyg** [comment]
> Awesome 🎆 Thanks for doing this!

no prob -- also I'm sure you're all aware of this, but it seems like the rats nest could be avoided by not exposing lower level crates publicly through higher level crates (like in holochain_zome_types, holochain_types, etc.), but rather expecting others to import the lower-level crates directly.

**Review by ThetaSinner — COMMENTED** [review]
@mattyg This is looking good! Two high level comments I have for this

- I don't see many removed dependencies, have you checked that it's possible to remove the problem dependencies after this change or is there more work to be done?
- I would like to see slightly fewer crates at some point and manage excluding code/deps with well-defined features. However, I actually think more crates is useful at the moment because it's easier to see what dependencies the code is actually using.

**Comment by mattyg** [comment]
> @mattyg This is looking good! Two high level comments I have for this
> 
>     * I don't see many removed dependencies, have you checked that it's possible to remove the problem dependencies after this change or is there more work to be done?

Yes I have a branch of holochain-client-rust that I just need to PR -- I'm able to avoid the problem dependencies after make these changes. 

The focus in this pr is removing `holochain_state`, `holochain_p2p` and `kitsune_p2p` from `holochain_conductor_api`, and pulling the nonce utility function out of `holochain_state`. Then I can avoid the problem imports to holochain-client-rust.

**Comment by ThetaSinner** [comment]
Super! I hadn't fully understood the end goal here then but that makes sense and I'm happy with this then

File: crates/kitsune_p2p/types/src/config.rs [hunk 1]
```diff
@@ -333,3 +339,324 @@ pub mod tuning_params_struct {
...
+                Ok(KitsuneP2pTx2Config { backend, use_proxy })
+            }
+            Some(TransportConfig::Quic { bind_to, .. }) => {
+                let bind_to = cnv_bind_to(bind_to);
+                Ok(KitsuneP2pTx2Config {
+                    backend: KitsuneP2pTx2Backend::Quic { bind_to },
+                    use_proxy: NoProxy,
+                })
+            }
+            */
```
Comment (jost-s) [inline crates/kitsune_p2p/types/src/config.rs:462]: Do we still need all these commented lines, as tx2 is being phased out?

File: crates/kitsune_p2p/types/src/config.rs [hunk 2]
```diff
@@ -333,3 +339,324 @@ pub mod tuning_params_struct {
...
+        /// We also need to define the sub-transport.
+        sub_transport: Box<TransportConfig>,
+
+        /// Determines whether we wish to:
+        /// - proxy through a remote
+        /// - be a proxy server for others
+        /// - be directly addressable, but not proxy for others
+        proxy_config: ProxyConfig,
+    },
+    */
```
Comment (jost-s) [inline crates/kitsune_p2p/types/src/config.rs:533]: see above

**Review by jost-s — APPROVED** [review]

File: crates/kitsune_p2p/types/src/config.rs (hunk 1 above)
Comment (jost-s) [reply crates/kitsune_p2p/types/src/config.rs:462]: We don't https://github.com/holochain/holochain/pull/2689/files#diff-1a8dd559c02e493cbcfa088e8da572bbe9ac95fc2ebcde414ccd942eab78661aL126-L161

File: crates/holochain/src/sweettest/sweet_network.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-//use kitsune_p2p::KitsuneP2pConfig;
```
[the remark this answers is not in the document]
Comment (maackle) [reply crates/holochain/src/sweettest/sweet_network.rs:1]: Created 3 years ago as block comments, I say let's remove it :) Too bad I can't suggest the deletion of an entire file on GH.

**Review by maackle — APPROVED** [review]
Thanks! Looks good. FYI I just discovered `git diff --color-moved` which was helpful to verify that nothing important was lost in the moves.

File: crates/hc_sandbox/Cargo.toml [hunk 4]
```diff
@@ -25,6 +25,7 @@ holochain_types = { path = "../holochain_types", version = "^0.3.0-beta-dev.16",
 holochain_websocket = { path = "../holochain_websocket", version = "^0.3.0-beta-dev.3"}
 holochain_p2p = { path = "../holochain_p2p", version = "^0.3.0-beta-dev.18", features = ["sqlite"] }
```
[the remark this answers is not in the document]
Comment (mattyg) [reply crates/hc_sandbox/Cargo.toml:26]: A bit more work -- its using these two utility functions from holochain_p2p: `agent_holo_to_kit` and `space_holo_to_kit` ( https://github.com/holochain/holochain/blob/78198e5ea611784e63aa3ebea5d67184ebef424c/crates/holochain_p2p/src/types.rs#L104 )

File: crates/holochain/src/sweettest/sweet_network.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattyg) [reply crates/holochain/src/sweettest/sweet_network.rs:1]: deleted the file

File: crates/holochain_conductor_api/Cargo.toml [hunk 5]
```diff
@@ -11,13 +11,13 @@ edition = "2021"
 [dependencies]
 directories = "2.0.2"
 derive_more = "0.99.3"
-kitsune_p2p = { version = "^0.3.0-beta-dev.16", path = "../kitsune_p2p/kitsune_p2p" }
+kitsune_p2p_types = { version = "^0.3.0-beta-dev.9", path = "../kitsune_p2p/types" }
+kitsune_p2p_bin_data = { version = "^0.3.0-beta-dev.6", path = "../kitsune_p2p/bin_data" }
```
[the remark this answers is not in the document]
Comment (mattyg) [reply crates/holochain_conductor_api/Cargo.toml:15]: yeah exactly

File: crates/holochain_state_types/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "holochain_state_types"
+version = "0.3.0-beta-dev.18"
+description = "Type for the holochain_state crate"
```
[the remark this answers is not in the document]
Comment (mattyg) [reply crates/holochain_state_types/Cargo.toml:4]: fixed

File: crates/kitsune_p2p/types/src/config.rs (hunk 1 above)
Comment (mattyg) [reply crates/kitsune_p2p/types/src/config.rs:462]: deleted

File: crates/holochain_nonce/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "holochain_nonce"
```
[the remark this answers is not in the document]
Comment (mattyg) [reply crates/holochain_nonce/Cargo.toml:2]: yeah my first attempt was to put it in holochain_util, but I ran into a circular dependency issue I think

File: crates/kitsune_p2p/types/src/config.rs (hunk 2 above)
Comment (mattyg) [reply crates/kitsune_p2p/types/src/config.rs:497]: deleted

**Comment by maackle** [comment]
This could also help with https://github.com/holochain/holochain/pull/2587

**Comment by maackle** [comment]
I just merged in develop and had to resolve conflicts in most Cargo.toml files since we just did a weekly release. I hope I didn't undo anything, you could have a look with `git diff 0edb1608e 3a508f1`

**Pull request merged**